### PR TITLE
Fiesta

### PR DESCRIPTION
**Pull request by felixdumais**
First commit with the 4 basic scripts - namely:

1. `ae_bundle_streamlines.py`;
2. `ae_find_threshold.py`;
3. `ae_generate_streamlines.py`;
4. `ae_train.py`.

2 configs file are also included in the `config` folder, needed for the scripts `ae_train.py` and `ae_find_threshold.py`.

### CODE REVIEW

**Comment by felixdumais** [comment]
@jhlegarreta I am not able to add you as a reviewer. Do I have the permissions ?

**Comment by jhlegarreta** [comment]
> @jhlegarreta I am not able to add you as a reviewer. Do I have the permissions ?

See if you can now.

**Comment by jhlegarreta** [comment]
Also, please rebase on `main` to fix the conflicts with the file `.github/labels.yml` and then push-force.

**Comment by felixdumais** [comment]
> > @jhlegarreta I am not able to add you as a reviewer. Do I have the permissions ?
> 
> See if you can now.

Yep, It works. Thanks

**Comment by felixdumais** [comment]
> Also, please rebase on `main` to fix the conflicts with the file `.github/labels.yml` and then push-force.

Yep, doing it RN

**Comment by felixdumais** [comment]
@jhlegarreta Also, let me know if I do weird stuff that you don't like in my git project management (commit naming, not enough information in my description, etc.)

I think you have more experience with that

**Comment by felixdumais** [comment]
> @jhlegarreta Also, let me know if I do weird stuff that you don't like in my git project management (commit naming, not enough information in my description, etc.)

I think you have more experience with that

File: .github/labels.yml [hunk 1]
```diff
@@ -47,7 +47,7 @@ type:packaging:
 type:infrastructure:
   - .gitignore
   - .github/*
-  - .pre-commit-config.yaml
+  - .pre-commit-train_config.yaml
```
Comment (jhlegarreta) [inline .github/labels.yml]: This should not have been changed. Any config file related to the method itself should be put be put under the `area:config:` labels. Adding this as a quick single, comment so that it is done in the first rebase while you fix conflicts. Will come up with a larger review.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+*.comet.config
+*.npy
+
+# Unit test / coverage reports
+.coverage
+
+log
+*/.ipynb_checkpoints
+
+setenv.sh
```
Comment (jhlegarreta) [inline .gitignore]: We do not have such a script in our code, right?

File: README.md [hunk 3]
```diff
@@ -8,6 +8,53 @@ J. H. Legarreta, M. Descoteaux, and P.-M. Jodoin. “PROCESSING OF TRACTOGRAPHY
...
+```sh
+   virtualenv tractolearn_env --python=python3.8
+   source tractolearn_env/bin/activate
+```
+
+They can be installed by executing:
+
+```sh
+   pip install -r requirements.txt
+   pip install -r requirements_dmri.txt
```
Comment (jhlegarreta) [inline README.md]: This has been deprecated: https://peps.python.org/pep-0518/

The above and some other related PEPs have introduced a more modern/robust way to specify the dependencies. I will come up with a new PR once this one gets merged to modernize that. In fact, the CI workflow file does already use such a modern convention.

File: README.md [hunk 4]
```diff
@@ -8,6 +8,53 @@ J. H. Legarreta, M. Descoteaux, and P.-M. Jodoin. “PROCESSING OF TRACTOGRAPHY
...
+```
+
+Torch not included. Tested with an NVIDIA RTX 3090 with:
```
Comment (jhlegarreta) [inline README.md]: Is there a reason for this? They should be in the `setup.cfg` version. Anyways I will specify them when transitioning to PEP0518.

File: README.md [hunk 5]
```diff
@@ -8,6 +8,53 @@ J. H. Legarreta, M. Descoteaux, and P.-M. Jodoin. “PROCESSING OF TRACTOGRAPHY
...
+
+To train deep learning models, you need to launch the script [ae_train.py](scripts/ae_train.py). This script takes a config file with all training parameters such as epochs, datasets path, etc. The most up-to-date config file is [config.yaml](configs/train_config.yaml). You can launch the training pipeline with the following command:
+```sh
+   ae_train.py train_config.yaml -vv
+```
+
+## Contributing
+
+## References
+
```
Comment (jhlegarreta) [inline README.md]: PR #14 should first be reviewed and merged, then this one rebased on `main` to fix the conflicts that will arise.

File: README.md [hunk 6]
```diff
@@ -8,6 +8,53 @@ J. H. Legarreta, M. Descoteaux, and P.-M. Jodoin. “PROCESSING OF TRACTOGRAPHY
...
+In order to execute experiments reporting to [Comet](https://www.comet.ml/site/), an `api_key` needs to be set as an environment variable named `COMETML`. You can write this command in you `.bashrc`
+
+```sh
+   export COMETML="api_key"
+```
+
+## Training models
+
+To train deep learning models, you need to launch the script [ae_train.py](scripts/ae_train.py). This script takes a config file with all training parameters such as epochs, datasets path, etc. The most up-to-date config file is [config.yaml](configs/train_config.yaml). You can launch the training pipeline with the following command:
```
Comment (jhlegarreta) [inline README.md]: I limit the line length to 79 characters: it helps reading/reviewing. There are tools out there breaking automatically for Markdown files, but we do not have many of such files, so breaking manually should be fine for now.

File: configs/find_threshold_config.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+model_type: "IncrFeatStridedConvFCUpsampReflectPadAE"
+model: "/home/fdumais/Desktop/Data/finta_rbx_atlas/models/best_model_contrastive_tractoinferno_hcp.pt"
```
Comment (jhlegarreta) [inline configs/find_threshold_config.yaml]: Avoid local paths. Use placeholders, relative paths or think of a way to make this work regardless of the file tree. Same applies to all instances below.

File: tractolearn/config/Experiment.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
```
Comment (jhlegarreta) [inline tractolearn/config/Experiment.py]: Use lower cases for any Python files:
https://peps.python.org/pep-0008/

File: tractolearn.egg-info/PKG-INFO [hunk 9]
```diff
@@ -0,0 +1,86 @@
+Metadata-Version: 2.1
```
Comment (jhlegarreta) [inline tractolearn.egg-info/PKG-INFO]: All these files should not have been committed. You may want to add a rule if it does not exist in the `.gitignore` file.

File: configs/train_config.yaml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+trk_path:
+hdf5_path: "/media/fdumais/My_Passport/Acuity/Data/TractoinfernoExtended/HDF5_QBx/data_tractoinferno_qbx_reduced.hdf5"
+pickle_path:
+rbx_classes: False
+
+# Reference anatomical filename (usually a t1.nii.gz or wm.nii.gz)
+ref_anat_fname: "/home/fdumais/Desktop/Data/finta_rbx_atlas/mni_masked.nii.gz"
+
+# Dataset name (e.g. `fibercup`, `ismrm2015_phantom`, etc.)
+dataset_name: mni_brain
```
Comment (jhlegarreta) [inline configs/train_config.yaml:13]: We may need to rethink this. By `dataset_name` I meant a dataset that is listed in some `anatomy/*.json` and `data/lut/*.json` to automatically e.g. color bundles with a prefixed colormap or obtain long names for datasets. `mni_brain` is not really indicative of any dataset out there; maybe you used it as a default/unimportant name.

File: requirements_dmri.txt [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#git+https://github.com/jhlegarreta/tractodata@main
+#git+ssh://git@github.com/fdumais/ismrm_2015_tractography_challenge_scoring.git
+git+ssh://git@github.com/fdumais/scilpy.git@master
```
Comment (jhlegarreta) [inline requirements_dmri.txt]: These should not point to our own forks (the first one's original source being my GitHub space). Unless they contain different commits that have not been merged upstream and have not been merged.

File: scripts/ae_train.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    loss_plot_fname = pjoin(trainer.run_dir, LoggerKeys.loss_plot_file_basename.value)
+    plot_loss_history(train_loss_data_fname, valid_loss_data_fname, loss_plot_fname)
+
+    logger.info(
+        "Loaded saved best model: loss {} at epoch: {}".format(
+            trainer.lowest_loss, lowest_loss_epoch
+        )
+    )
+
+    logger.info("Finished experiment.")
```
Comment (jhlegarreta) [inline scripts/ae_train.py:191]: In a second pass, we will be willing to refactor this entire block into methods that can be re-used, the `main()` virtually containing a few method calls and being as short as possible. Same applies to any other script.

File: tractolearn/generative/filter_generated_points.py [hunk 13]
```diff
@@ -0,0 +1,657 @@
+# -*- coding: utf-8 -*-
```
Comment (jhlegarreta) [inline tractolearn/generative/filter_generated_points.py]: I'd rename the module:
- A module should not contain an action verb, as opposed to a script that actually performs some processing.
- `points` is maybe a vague term.

This module contains methods that belong to some other module, maybe `analysis`.  So the `filter` term is also vague here. This, essentially, is performing the sample data discrimination.

File: tractolearn/generative/filter_generated_points.py [hunk 14]
```diff
@@ -0,0 +1,657 @@
...
+    ids : tuple
+        Filtered sft.
+        Ids of the streamlines passing through the mask.
+    """
+    line_based_indices = []
+    if filter_type in ["any", "all"]:
+        line_based_indices = streamlines_in_mask(sft, mask, all_in=filter_type == "all")
+    else:
+        sft.to_vox()
+        sft.to_corner()
```
Comment (jhlegarreta) [inline tractolearn/generative/filter_generated_points.py]: Should be put into a function to always do the same thing, in the same manner, and to avoid forgetting one of the two. Same applies to undoing it.

File: tractolearn/learning/Datasets.py [hunk 15]
```diff
@@ -0,0 +1,452 @@
+# -*- coding: utf-8 -*-
```
Comment (jhlegarreta) [inline tractolearn/learning/Datasets.py]: Same comment as above concerning the module name. Also use singular nouns. Different classes may need to go into separate files.

File: tractolearn/utils/Timer.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
```
Comment (jhlegarreta) [inline tractolearn/utils/Timer.py]: Same comment as above concerning the module naming.

File: tractolearn/tractoio/utils.py [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+                np.newaxis,
+            ]
+            for i in range(len(flipped_streamlines))
+        ]
+    )
+
+    if streamline_class_name == "plausible":
+        streamlines_class = 0
+    elif streamline_class_name == "implausible":
+        streamlines_class = 100
```
Comment (jhlegarreta) [inline tractolearn/tractoio/utils.py:239]: We'll need to improve this so that it is not hard-coded.

The block may best fit into a dedicated method.

File: tractolearn/utils/losses.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+    """Attract pairs of latent vectors of the same class, repulse pairs of different classes
+
+    This is the contrastive loss as defined by Hadsell, Chopra and LeCun, 2006. However, in their paper, they don't
+    use class information.
+
+    Parameters:
+        z       Tensor of size (num_pos_pairs * 2 + num_neg_pairs * 2, latent_size). This is the batch format output by
+                ContrastiveDataset.
+        margin  The margin hyperparameter
+    """
```
Comment (jhlegarreta) [inline tractolearn/utils/losses.py:51]: Use the NumPy convention for documenting.

File: tractolearn/utils/utils.py [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+        root_output_path = out_path
+    run_dir = Path(pjoin(root_output_path, run_name))
+    run_dir.mkdir()
+    return run_dir
+
+
+def generate_uuid():
+
+    eventid = datetime.datetime.now().strftime("%Y%m%d-%H%M%S-%f-") + str(uuid4())
+    return eventid
```
Comment (jhlegarreta) [inline tractolearn/utils/utils.py:25]: With time, I am less inclined to use this method to name folders. It's OK for now, but converging towards a more robust/readable/BIDS-friendly solution would be desirable eventually.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
A few comments.

Some are thoughts to be addressed in the future.

**Comment by jhlegarreta** [comment]
> @jhlegarreta Also, let me know if I do weird stuff that you don't like in my git project management (commit naming, not enough information in my description, etc.)
> I think you have more experience with that

Avoid adding merge commits: rebase on `main`, and push-force. Do not add "comment fix/address"-like commits: amend the corresponding commits. Care must be taken to do this properly, I know, but keeping a clean history pays off down the road.

File: scripts/ae_find_thresholds.py [hunk 20]
```diff
@@ -0,0 +1,431 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Script for the inference of FINTA
```
Comment (jhlegarreta) [inline scripts/ae_find_thresholds.py]: Does not correspond to what the script does.

File: scripts/ae_generate_streamlines.py [hunk 21]
```diff
@@ -0,0 +1,532 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Script for the inference of FINTA
```
Comment (jhlegarreta) [inline scripts/ae_generate_streamlines.py]: Does not correspond to what the script does.

File: scripts/ae_bundle_streamlines.py [hunk 22]
```diff
@@ -0,0 +1,438 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Script for the inference of FINTA
```
Comment (jhlegarreta) [inline scripts/ae_bundle_streamlines.py]: Does not correspond to what the script does.

File: scripts/ae_generate_streamlines.py [hunk 23]
```diff
@@ -0,0 +1,532 @@
...
+    )
+    parser.add_argument(
+        "--in_deformation",
+        help="Path to the file containing a deformation field [ *.nii/.nii.gz ].",
+    )
+
+    parser.add_argument(
+        "--white_matter_config",
+        help="Config file [ *.json ] used to determine the type of WM filtering ( WM mask | thresholded FA ).",
+    )
```
Comment (jhlegarreta) [inline scripts/ae_generate_streamlines.py:212]: So the paradigm of the parameters here is different wrt to e.g. the threshold finding script, where a single config file is used. We will eventually need to be consistent with this.

File: tractolearn/generative/filter_generated_points.py (hunk 13 above)
Comment (felixdumais) [reply tractolearn/generative/filter_generated_points.py]: I moved the file in the filtering folder and called it: `streamline_space_filtering.py`

File: scripts/ae_generate_streamlines.py (hunk 23 above)
Comment (felixdumais) [reply scripts/ae_generate_streamlines.py:212]: Yep, I guess we could pass config file for the `ae_find_thresholds.py` script. This was mandatory for the `ae_generate_streamlines.py` script as fsl WM mask almost always output 0 value around the fornix/AC/PC regions

File: README.md (hunk 3 above)
Comment (felixdumais) [reply README.md]: Do you want me to do something about it ? You want to use the `pyproject.toml` approach I guess

Comment (jhlegarreta) [reply README.md]: We will not be able to test that the PR is passing if we do not transition to it. But maybe it can be done in a future PR.

Comment (felixdumais) [reply README.md]: Ok good

File: configs/find_threshold_config.yaml (hunk 7 above)
Comment (felixdumais) [reply configs/find_threshold_config.yaml]: I modified the scripts to pass to variable as parameters

File: .gitignore (hunk 2 above)
Comment (jhlegarreta) [reply .gitignore]: Not sure if I still see this. If this is local to a user, it should generally not be added.

Comment (felixdumais) [reply .gitignore]: It should be removed now

File: README.md (hunk 4 above)
Comment (felixdumais) [reply README.md]: Added in the requirements.txt

File: README.md (hunk 6 above)
Comment (felixdumais) [reply README.md]: Refactored the readme

File: tractolearn/utils/Timer.py (hunk 16 above)
Comment (felixdumais) [reply tractolearn/utils/Timer.py]: Done

File: README.md (hunk 3 above)
Comment (felixdumais) [reply README.md]: Done

File: .gitignore (hunk 2 above)
Comment (felixdumais) [reply .gitignore]: Done

File: README.md (hunk 4 above)
Comment (felixdumais) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (felixdumais) [reply README.md]: To add in a further PR

File: README.md (hunk 6 above)
Comment (felixdumais) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (jhlegarreta) [reply README.md]: No need to add anything manually; rebasing on `main` should be enough.

Comment (felixdumais) [reply README.md]: Done

File: configs/find_threshold_config.yaml (hunk 7 above)
Comment (felixdumais) [reply configs/find_threshold_config.yaml]: Done

File: tractolearn/config/Experiment.py (hunk 8 above)
Comment (felixdumais) [reply tractolearn/config/Experiment.py]: Done

File: tractolearn.egg-info/PKG-INFO (hunk 9 above)
Comment (felixdumais) [reply tractolearn.egg-info/PKG-INFO]: Done

File: scripts/ae_train.py (hunk 12 above)
Comment (felixdumais) [reply scripts/ae_train.py:191]: To do in another PR

File: tractolearn/generative/filter_generated_points.py (hunk 13 above)
Comment (felixdumais) [reply tractolearn/generative/filter_generated_points.py]: Done

File: tractolearn/generative/filter_generated_points.py (hunk 14 above)
Comment (felixdumais) [reply tractolearn/generative/filter_generated_points.py]: Done

File: tractolearn/learning/Datasets.py (hunk 15 above)
Comment (felixdumais) [reply tractolearn/learning/Datasets.py]: Done

File: tractolearn/utils/utils.py (hunk 19 above)
Comment (felixdumais) [reply tractolearn/utils/utils.py:25]: To do in another PR

File: scripts/ae_find_thresholds.py (hunk 20 above)
Comment (felixdumais) [reply scripts/ae_find_thresholds.py]: Done

File: scripts/ae_generate_streamlines.py (hunk 21 above)
Comment (felixdumais) [reply scripts/ae_generate_streamlines.py]: Done

File: scripts/ae_bundle_streamlines.py (hunk 22 above)
Comment (felixdumais) [reply scripts/ae_bundle_streamlines.py]: Done

File: scripts/ae_generate_streamlines.py (hunk 23 above)
Comment (felixdumais) [reply scripts/ae_generate_streamlines.py:212]: Maybe to do in another PR ?

File: tractolearn/utils/losses.py (hunk 18 above)
Comment (felixdumais) [reply tractolearn/utils/losses.py:51]: Done

File: .github/labels.yml [hunk 24]
```diff
@@ -47,7 +48,6 @@ type:packaging:
 type:infrastructure:
   - .gitignore
   - .github/**/*
-  - .pre-commit-config.yaml
```
Comment (jhlegarreta) [inline .github/labels.yml]: Why is this change made?

File: .github/workflows/test_package.yml [hunk 25]
```diff
@@ -2,9 +2,9 @@ name: test, package
```
Comment (jhlegarreta) [inline .github/workflows/test_package.yml]: I do not see why this file has been changed. This is unrelated to the content of the code/package itself.

File: .github/scripts/pr_title_regex_labeler.py [hunk 26]
```diff
@@ -30,8 +30,7 @@
 # Add the word boundary to build the title regex
 word_boundary = r"\b"
 title_to_labels_regex = [
-    (word_boundary + key + word_boundary, val)
-    for key, val in title_to_labels.items()
+    (word_boundary + key + word_boundary, val) for key, val in title_to_labels.items()
```
Comment (jhlegarreta) [inline .github/scripts/pr_title_regex_labeler.py]: Same as above.

File: README.md [hunk 27]
```diff
@@ -10,24 +10,30 @@ Tractography learning.
 ## How to cite
```
Comment (jhlegarreta) [inline README.md]: Not sure why this has changed the format; if a linter was applied, I would do that in a separate PR. Not sure either which was the linter used; if a particular configuration is used, it should be shared, and there is a PR in progress to apply linters (at least to the code itself).

File: .github/scripts/issue_title_regex_labeler.py [hunk 28]
```diff
@@ -29,8 +29,7 @@
 # Add the word boundary to build the title regex
 word_boundary = r"\b"
 title_to_labels_regex = [
-    (word_boundary + key + word_boundary, val)
-    for key, val in title_to_labels.items()
+    (word_boundary + key + word_boundary, val) for key, val in title_to_labels.items()
```
Comment (jhlegarreta) [inline .github/scripts/issue_title_regex_labeler.py]: Although it is fine to pass the linter in here, I'd leave that for a separate PR to separate things clearly.

File: tractolearn/config/experiment.py [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+    BUNDLE_KEYS = "bundle_keys"
+    DATA_IN_MEMORY = "data_in_memory"
+    NUM_WORKERS = "num_workers"
+    DISTANCE_FUNCTION = "distance_function"
+    TO_SWAP = "to_swap"
+
+
+class ThresholdTestKeys(
+    Enum
+):  # TODO do no inherit from Enum to avoid always using .value
```
Comment (jhlegarreta) [inline tractolearn/config/experiment.py]: Does this comment still hold? Should the other classes inherit from `Enum`?

File: tractolearn/config/experiment.py [hunk 30]
```diff
@@ -0,0 +1,209 @@
...
+        hdf5_path: str,
+        bundle_keys_path: str,
+        output_path: str,
+        ref_anat_fname: str,
+    ):
+        """
+
+        Args:
+            config: Config file
+        """
```
Comment (jhlegarreta) [inline tractolearn/config/experiment.py:139]: Please use `NumPy` documentation convention. Have not exhaustively verified all files.

File: tractolearn/utils/losses.py [hunk 31]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def loss_triplet_classes(z, margin, metric="l2", swap=False):
+    """Anchors attract positive values and repulse negatives values
+
+    Parameters:
+        z       Tensor of size (num_pos_pairs * 2 + num_neg_pairs * 2, latent_size). This is the batch format output by
+                TripletDataset.
+        margin  The margin hyperparameter
+    """
```
Comment (jhlegarreta) [inline tractolearn/utils/losses.py]: Please use NumPy convention for documentation.

File: configs/train_config.yaml (hunk 10 above)
Comment (felixdumais) [reply configs/train_config.yaml:13]: To do in another PR

**Comment by jhlegarreta** [comment]
Force pushed. Improvements to be made once this gets merged.

**Review by jhlegarreta — APPROVED** [review]
Thanks for this Félix. As discussed, the CI not passing is expected. A separate PR will be made to adapt the necessary package files.

Merging.

File: .github/labels.yml (hunk 24 above)
Comment (jhlegarreta) [reply .github/labels.yml]: Reverted.

File: .github/workflows/test_package.yml (hunk 25 above)
Comment (jhlegarreta) [reply .github/workflows/test_package.yml]: Reverted.

File: .github/scripts/pr_title_regex_labeler.py (hunk 26 above)
Comment (jhlegarreta) [reply .github/scripts/pr_title_regex_labeler.py]: Reverted.

File: .github/scripts/issue_title_regex_labeler.py (hunk 28 above)
Comment (jhlegarreta) [reply .github/scripts/issue_title_regex_labeler.py]: Reverted.

File: .github/labels.yml (hunk 24 above)
Comment (jhlegarreta) [reply .github/labels.yml]: Reverted.

File: README.md (hunk 27 above)
Comment (jhlegarreta) [reply README.md]: Reverted.

File: tractolearn/config/experiment.py (hunk 30 above)
Comment (jhlegarreta) [reply tractolearn/config/experiment.py:139]: Done.

File: tractolearn/config/experiment.py (hunk 29 above)
Comment (felixdumais) [reply tractolearn/config/experiment.py]: Yes, I think we should avoid inheritance for this usage. It is much easier to read if we just define a simple class.

**Review by jhlegarreta — APPROVED** [review]
:+1: Merging.
